example-crates/wallet_esplora_async/src/main.rs

```diff
@@ -24,62 +32,93 @@ async fn main() -> Result<(), anyhow::Error> {
-
-    print!("Syncing...");
-    let client =
-        esplora_client::Builder::new("https://blockstream.info/testnet/api").build_async()?;
-
-    let prev_tip = wallet.latest_checkpoint();
-    let keychain_spks = wallet
-        .all_unbounded_spk_iters()
-        .into_iter()
-        .map(|(k, k_spks)| {
-            let mut once = Some(());
-            let mut stdout = std::io::stdout();
-            let k_spks = k_spks
-                .inspect(move |(spk_i, _)| match once.take() {
-                    Some(_) => print!("\nScanning keychain [{:?}]", k),
-                    None => print!(" {:<3}", spk_i),
-                })
-                .inspect(move |_| stdout.flush().expect("must flush"));
-            (k, k_spks)
-        })
-        .collect();
-    let (update_graph, last_active_indices) = client
-        .full_scan(keychain_spks, STOP_GAP, PARALLEL_REQUESTS)
-        .await?;
-    let missing_heights = update_graph.missing_heights(wallet.local_chain());
-    let chain_update = client.update_local_chain(prev_tip, missing_heights).await?;
-    let update = Update {
-        last_active_indices,
-        graph: update_graph,
-        chain: Some(chain_update),
-    };
+    info!("Wallet balance: {} sats", balance.total());
+
+    let client = esplora_client::Builder::new("http://signet.bitcoindevkit.net").build_async()?;
+
+    let (update, cmd) = match cmd.map(|c| c.as_str()) {
+        Some(cmd) if cmd == "fullscan" => {
+            info!("Start full scan...");
+            // 1. get data required to do a wallet full_scan
+            let request = wallet.full_scan_request();
+            // 2. full scan to discover wallet transactions
+            let (graph_update, last_active_indices) = client
+                .full_scan(request.spks_by_keychain, STOP_GAP, PARALLEL_REQUESTS)
+                .await?;
+            // 3. from wallet transactions determine missing blockchain heights
+            let missing_heights = graph_update.missing_heights(wallet.local_chain());
+            // 4. get blockchain update from original request checkpoint and missing heights
+            let chain_update = client
+                .update_local_chain(request.checkpoint, missing_heights)
+                .await?;
+            // 5. create wallet update
+            Ok((
+                Update {
+                    last_active_indices,
+                    graph: graph_update,
+                    chain: Some(chain_update),
+                },
+                cmd,
+            ))
+        }
+        Some(cmd) if cmd == "sync" => {
+            info!("Start sync...");
+            // 1. get data required to do a wallet sync, if also syncing previously used addresses set unused_spks_only = false
+            let request = wallet.sync_request(true);
+            // 2. sync unused wallet spks (addresses), unconfirmed tx, and utxos
+            let graph_update = client
+                .sync(
+                    request.spks,
```

Comment:
I'm guessing the point of this PR is to make the esplora/electrum apis take the request as the argument here?

Reply:
Updated to use the request structs.